File: src/main/java/org/veupathdb/lib/blast/field/MTMode.java
package org.veupathdb.lib.blast.field;

import com.fasterxml.jackson.databind.JsonNode;
import org.veupathdb.lib.blast.util.DefaultingJSONValue;
import org.veupathdb.lib.blast.util.JSONConstructor;

public enum MTMode implements DefaultingJSONValue
{
  SplitByDBVolumes,
  SplitByQueries;

  public static final MTMode DefaultMTMode = SplitByDBVolumes;

  public int getValue() {
    return ordinal();
  }

  @Override
  public boolean isDefault() {
    return this == DefaultMTMode;
  }

  @Override
  public JsonNode toJSON() {
    return JSONConstructor.newInt(ordinal());
  }

  public static MTMode fromIntValue(int value) {
    for (var val : values())
      if (val.getValue() == value)
        return val;

    throw new IllegalArgumentException(
      "Illegal mt_mode value \"" + value + "\", must be one of 0 or 1"
    );
  }

  public static MTMode fromJSON(JsonNode js) {
    return fromIntValue(js.intValue());
  }
}

File: src/main/java/org/veupathdb/lib/blast/field/MTMode.kt
package org.veupathdb.lib.blast.field

import org.veupathdb.lib.blast.util.DefaultingJSONValue
import com.fasterxml.jackson.annotation.JsonValue
import com.fasterxml.jackson.databind.JsonNode
import org.veupathdb.lib.blast.util.JSONConstructor
import com.fasterxml.jackson.annotation.JsonCreator
import java.lang.IllegalArgumentException

enum class MTMode : DefaultingJSONValue {
  SplitByDBVolumes, SplitByQueries;

  fun getValue(): Int {
    return ordinal
  }

  override val isDefault
    get() = this == DefaultMTMode

  @JsonValue
  override fun toJSON(): JsonNode {
    return JSONConstructor.newInt(ordinal)
  }

  companion object {
    @JvmStatic
    val DefaultMTMode = SplitByDBVolumes

    @JvmStatic
    fun fromIntValue(value: Int): MTMode {
      for (`val` in values()) if (`val`.getValue() == value) return `val`
      throw IllegalArgumentException(
        "Illegal mt_mode value \"$value\", must be one of 0 or 1"
      )
    }

    @JvmStatic
    @JsonCreator
    fun fromJSON(js: JsonNode): MTMode {
      return fromIntValue(js.intValue())
    }
  }
}